applications/tari_testnet_miner/src/miner.rs

```diff
@@ -0,0 +1,154 @@
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use chrono::{DateTime, Duration, Utc};
+use derive_error::Error;
+use digest::Digest;
+use rand::OsRng;
+use tari_core::{
+    blocks::{Block, BlockHeader},
+    consensus::ConsensusRules,
+    proof_of_work::*,
+    tari_amount::MicroTari,
+    transaction::*,
+    types::*,
+};
+use tari_crypto::{
+    commitment::HomomorphicCommitmentFactory,
+    keys::{PublicKey as PK, SecretKey},
+    range_proof::RangeProofService,
+};
+use tari_utilities::byte_array::ByteArray;
+
+/// TestNet Miner
+pub struct Miner {
+    // current working block
+    block: Option<Block>,
+    // prev block's solved difficulty, TODO check this
+    difficulty: Difficulty,
+    // current consensus rules
+    rules: ConsensusRules,
+}
+
+#[derive(Clone, Debug, PartialEq, Error)]
+pub enum MinerError {
+    // Could not construct the coinbase utxo and kernel for the block
+    CoinbaseError,
+    // No block provided to mine
+    MissingBlock,
+}
+
+impl Miner {
+    /// Create a new empty miner
+    pub fn new(rules: ConsensusRules) -> Miner {
+        Miner {
+            block: None,
+            difficulty: Difficulty::min(),
+            rules,
+        }
+    }
+
+    /// Add a block to be processed.  This function also add the coinbase tx
+    pub fn add_block(&mut self, mut block: Block, coinbase_key: PrivateKey) -> Result<(), MinerError> {
+        let mut rng = rand::OsRng::new().unwrap();
+        let coinbase_value =
+            block.calculate_coinbase_and_fees(self.rules.emission_schedule().block_reward(block.header.height));
+        let new_commitment = COMMITMENT_FACTORY.commit(&coinbase_key, &coinbase_value.into());
+        let rr = PROVER
+            .construct_proof(&coinbase_key, coinbase_value.into())
+            .map_err(|_| MinerError::CoinbaseError)?;
+        let coinbase = TransactionOutput {
+            commitment: new_commitment,
+            features: OutputFeatures::create_coinbase(block.header.height, &self.rules),
+            proof: RangeProof::from_bytes(&rr).unwrap(),
+        };
+        let excess = COMMITMENT_FACTORY.commit(&coinbase_key, &(MicroTari(0)).into());
+        let nonce = PrivateKey::random(&mut rng);
+        let challenge = Miner::get_challenge(&PublicKey::from_secret_key(&nonce));
+        let sig = Signature::sign(coinbase_key, nonce, &challenge).map_err(|_| MinerError::CoinbaseError)?;
+        let kernel = TransactionKernel {
+            features: KernelFeatures::empty(),
+            fee: 0.into(),
+            lock_height: 0,
+            excess,
+            excess_sig: sig,
+            meta_info: None,
+            linked_kernel: None,
+        };
+        block.body.outputs.push(coinbase);
```

Comment:
This a bug in `AggregateBody` -- inputs and outputs should _not_ be public.
You _must_ use `add_output` here, otherwise the block won't be sorted properly.